### PR TITLE
kNN brute force optimizations

### PR DESCRIPTION
**Pull request by SmirnovEgorRu**
For Airline (100k rows): before 40053 ms, now 312ms. **128x speedup**

### CODE REVIEW

**Comment by SmirnovEgorRu** [comment]
I checked accuracy for mnist dataset for k=5 with Scikit-learn.
Indexes, distances and labels are the same.

File: algorithms/kernel/k_nearest_neighbors/bf_knn_impl.i [hunk 1]
```diff
@@ -94,107 +72,203 @@ public:
...
+
+        ReadRows<FPType, cpu> inDataRows(const_cast<NumericTable *>(testTable), i1, i2 - i1);
+        DAAL_CHECK_BLOCK_STATUS(inDataRows);
+        const FPType * const testData = inDataRows.get();
+
+        const size_t outBlockSize = trainBlockSize;
+        const size_t outRows      = nTrain;
+        const size_t nOutBlocks   = outRows / outBlockSize + (outRows % outBlockSize > 0);
+
+        for (size_t outBlock = 0; outBlock < nOutBlocks; outBlock++)
```
Comment (PetrovKP) [inline algorithms/kernel/k_nearest_neighbors/bf_knn_impl.i:151]: Why not nest the parallel for?
If nTrain >> nTest  parallelism will be ineffective

Comment (PetrovKP) [reply algorithms/kernel/k_nearest_neighbors/bf_knn_impl.i:151]: But in the general case, this should still give a win.

File: algorithms/kernel/k_nearest_neighbors/bf_knn_impl.i [hunk 2]
```diff
@@ -94,107 +72,203 @@ public:
...
         if (resultsToCompute & computeIndicesOfNeightbors)
         {
-            for (size_t i = 0; i < blockSize; ++i)
-            {
-                for (size_t j = 0; j < k; ++j)
-                {
-                    indices[(i + startTestIdx) * k + j] = tmpIndices[i * nTrain + j];
-                }
-            }
+            daal::internal::WriteRows<int, cpu> indexesBlock(indicesTable, startTestIdx, blockSize);
```
Comment (PetrovKP) [inline algorithms/kernel/k_nearest_neighbors/bf_knn_impl.i:185]: There may be a conversion while reading. This is not necessary here
```suggestion
            daal::internal::WriteOnlyRows<int, cpu> indexesBlock(indicesTable, startTestIdx, blockSize);
```

File: algorithms/kernel/k_nearest_neighbors/bf_knn_impl.i [hunk 3]
```diff
@@ -94,107 +72,203 @@ public:
...
         if (resultsToCompute & computeDistances)
         {
-            for (size_t i = 0; i < blockSize; ++i)
-            {
-                for (size_t j = 0; j < k; ++j)
-                {
-                    distances[(i + startTestIdx) * k + j] = tmpDistances[i * nTrain + j];
-                }
-            }
+            daal::internal::WriteRows<FPType, cpu> distancesBlock(distancesTable, startTestIdx, blockSize);
```
Comment (PetrovKP) [inline algorithms/kernel/k_nearest_neighbors/bf_knn_impl.i:195]: Same comment
```suggestion
            daal::internal::WriteOnlyRows<FPType, cpu> distancesBlock(distancesTable, startTestIdx, blockSize);
```

**Review by PetrovKP — APPROVED** [review]

File: algorithms/kernel/k_nearest_neighbors/bf_knn_impl.i [hunk 4]
```diff
@@ -94,107 +72,203 @@ public:
             DAAL_CHECK_MALLOC(trainLabel);
         }
 
-        daal::algorithms::internal::EuclideanDistances<FPType, cpu> euclDist(
-            *testTable, *trainTable, !(resultsToCompute & computeDistances) && !(voteWeights & VoteWeights::voteDistance));
+        daal::algorithms::internal::EuclideanDistances<FPType, cpu> euclDist(*testTable, *trainTable, true);
```
Comment (Alexsandruss) [inline algorithms/kernel/k_nearest_neighbors/bf_knn_impl.i:75]: Why compute distances without sqrt (set squared=false)?

**Review by Alexsandruss — APPROVED** [review]

File: algorithms/kernel/k_nearest_neighbors/bf_knn_impl.i (hunk 4 above)
Comment (SmirnovEgorRu) [reply algorithms/kernel/k_nearest_neighbors/bf_knn_impl.i:75]: Computed in another place

**Pull request merged**